lib/nolibc: Update musl-imported headers for signals

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`arm64`, `x86_64`]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series updates the `musl-imported` headers for signals:
* Import `feature.h` from musl to simplify management of FTMs (https://lwn.net/Articles/590381/)
* Replace `signal.h` with a clean copy from musl to simplify manitenance
* Import architecture-specifi headers for `arm64` and `x86_64`
* Import additional definitions to resolve dependencies

michpappas committed:
lib/nolibc: Import architecture signal definitions from musl

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: musl/arch/[aarch64,x86_64]/bits/signal.h

Checkpatch-Ignore: LOGICAL_CONTINUATIONS
Checkpatch-Ignore: MACRO_ARG_REUSE
Checkpatch-Ignore: UNSPECIFIED_INT
Checkpatch-Ignore: LEADING_SPACE
Checkpatch-Ignore: LINE_SPACING
Checkpatch-Ignore: BIT_MACRO
Checkpatch-Ignore: SPACING
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Import _POSIX_SIGQUEUE_MAX from musl

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: /home/mpp/devel/musl/include/limits.h

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Import `pthread_t` & `pthread_attr_t` from musl

Import definitions of `pthread_d` and `pthread_addr_t`
from musl, with minor changes to adapt for nolibc's
`sharedefs.h`.

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: musl/include/alltypes.h.in

Checkpatch-Ignore: POINTER_LOCATION
Checkpatch-Ignore: LONG_LINES
Checkpatch-Ignore: VOLATILE
Checkpatch-Ignore: SPACING
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Import `signal.h` from musl

This replaces `signal.h` with a clean copy from
musl to simplify manitenance.

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: musl/include/alltypes.h.in

Checkpatch-Ignore: LOGICAL_CONTINUATIONS
Checkpatch-Ignore: FUNCTION_ARGUMENTS
Checkpatch-Ignore: UNSPECIFIED_INT
Checkpatch-Ignore: LEADING_SPACE
Checkpatch-Ignore: LONG_LINE
Checkpatch-Ignore: BIT_MACRO
Checkpatch-Ignore: SPACING
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Adapt musl imported `signal.h` to nolibc

The replaced values of SIGRTMIN and SIGRTMAX are effectively
the return values of musl's __libc_current_sigrtmin() and
__libc_current_sigrtmax() defined in musl/src/signal/sigrtmin.c
and musl/src/signal/sigrtmax.c

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Rename musl-imported/arch/aarch64

Rename `musl-imported/arch/aarch64` to `musl-imported/arch/arm64`.
to fix build errors caused by an incorrect include path generated
relative to $(ARCH).

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Add _GNU_SOURCE to strsignal.c

After updating signal.h from musl strsignal.c requires
_GNU_SOURCE for _NSIG.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/nolibc: Import features.h from musl

This simplifies use of feature test macros.

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: musl/include/alltypes.h.in

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

csvancea commented on lib/nolibc/musl-imported/include/signal.h:
```diff
@@ -83,63 +178,124 @@ struct sigaction {
+#define SYS_USER_DISPATCH 2
+#endif
 
+#ifdef _GNU_SOURCE
 typedef void (*sighandler_t)(int);
```
`lib/uksignal/signal.c` does not define `_GNU_SOURCE`, which leads to `sighandler_t` being undefined here: https://github.com/michpappas/unikraft/blob/1d52b24e408c0918d967f643494668d1991492bc/lib/uksignal/signal.c#L87

Of course this is tackled in #1248, but is it expected to have build failures in-between this PR and #1248?

michpappas replied:
Thanks, fixed

John-Ted commented:
Can you rebase this on the latest main?

michpappas committed:
lib/uksignal: Set _GNU_SOURCE to preserve definition of sighandler_t

Adding feature.h in nolibc requires that we set _GNU_SOURCE to preserve
the definition of sighandler_t in signal.c

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Rebased from `staging`. Notice that the checkpatch errors are due to:
* Tags for tracability that checkpatch confuses for invalid SPDX tags
* Non-confirming syntax of musl (we don't adapt to our convention)

andreittr commented on lib/nolibc/musl-imported/src/string/strsignal.c:
```diff
@@ -2,6 +2,7 @@
 
 #include <uk/essentials.h>
 #include <string.h>
+#define _GNU_SOURCE
```
Hate to be nitpicky, but I think we should define any feature test macros before including any header files. The `feature_test_macros(7)` manpage says as much in the first note: "In order to be effective, a feature test macro must be defined before including any header files.". Or is there a very compelling reason why we'd want only `<signal.h>` to see this macro?

(We know `<uk/essentials.h>` doesn't use any feature macros, so we can maybe ignore that, but `<string.h>` is outside of our control (technically it isn't since we're inside nolibc, but fewer assumptions make for cleaner code)).

michpappas replied:
Not nitpicky at all, as you pointed out FTMs must be defined first. No idea why I did it that way back then, but I testing the proper way didn't show any problems, so I will push an update. Thanks for catching that.

andreittr reviewed (approved):
Looks good, thanks for the work!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Approved-by: Andrei Tatar <andrei@unikraft.io>
